### DISCUSSION TITLE
Displaying fields from a relationship

### DISCUSSION THREAD

**Discussion by haztakki** [laravel/nova-issues#4423] [Q&A] [answered]
Hey,
Simple question but I can't seem to find what I'm looking for in the docs. How can I display columns from a relationship in the table listing? I'm probably missing it. If you could point me in the right direction, that would be superb. Just looking for links or a bit of information to get me going. Overall Nova is such a breeze to work with.

https://nova.laravel.com/docs/4.0/resources/fields.html
https://nova.laravel.com/docs/4.0/resources/relationships.html

<img width="721" alt="Screenshot 2022-06-14 at 22 48 22" src="https://user-images.githubusercontent.com/20573863/173675941-3bd79dd9-6ead-4bd8-9816-d19d9530a571.png">

Let's say for example, I wanted to display a user's street address in the listing here (see above). The street column is available on the addresses relationship.

Taking this a step further, is there a way to do more advanced stuff, like take the above example and have it display the user's primary address from the addresses relationship? The primary address is also a column on the relationship.

Lastly, is there a way to grab data (that we know of) and display it in a column? We currently use the Code field to display the whole thing in the detailed view, though it would be great if we could display some the stuff in the listing also. The only thing that springs to mind is having multiple accessors on the model.

<img width="1076" alt="Screenshot 2022-06-14 at 22 51 07" src="https://user-images.githubusercontent.com/20573863/173676554-6cfb2203-fb6c-4262-9286-9103061fb1cc.png">

**Comment by juse-less** [comment, marked as answer]
I can't remember if I've tested it on relationships, but, I'm showing 'nested properties' using dot-notation.
Basically, say, my user has a JSON column of values, or I'm using casts to build a property that is an array, then I can refer to one of the nested properties using dot-notation.

For your use-case, maybe dot-notation works on referencing relationships as well?
```php
Text::make(__('Primary Street Address'), attribute: 'addresses.primary.street')
```

Other than that, what comes to mind directly is computed fields.
```php
Text::make(__('Primary Street Address'), function (User $user): string {
    // Replace with what makes sense. I'm assuming your 'addresses' aren't keyed like this,
    //   but for simplicity in this example.
    return $user->addresses->get('primary')?->street;
})
```

**Reply by robbedekien** [reply]
Just a heads up, I was getting sql errors when trying to make these columns sortable. 

If you want a sortable column for a nested relationship you need to add select statements to the indexQuery method. Then you can make the Text field sortable.

```
public static function indexQuery(NovaRequest $request, $query)
    {
        // Define extra selects for showing relation fields
        $query
            ->join('addresses', 'addresses.id', '=', 'contacts.address_id')
            ->select([
                'contacts.*',
                'addresses.street_name as address_street_name',
            ]);

        if (empty($request->get('orderBy'))) {
            $query->getQuery()->orders = [];
            return $query->orderBy(key(static::$sort), reset(static::$sort));
        }
        
        return $query;
    }
```

Then you can define your Text field like this
```
      Text::make(__('Street name'), 'address_street_name')
          ->showOnPreview()
          ->sortable()
```

**Comment by haztakki** [comment]
Oh wow. Magic! The dot notation works great for both relationships and JSON. Thanks @juse-less!

_Example 1_

```
Text::make(name: 'Street', attribute: 'addresses.0.street'),
```

_Example 2_

```
            Text::make(name: 'Street', attribute: function ($user) {
                return $user->addresses->where('primary', true)->first()?->street;
            }),
```

_Example 3_

```
            Text::make(name: 'Email', attribute: 'data.contact.email'),

            Text::make(name: 'Phone', attribute: 'data.contact.phone'),
```

**Reply by juse-less** [reply]
Sweet!
Glad I could help. 🙂
